### PR TITLE
Issue519 : camera page

### PR DESCRIPTION
**Pull request by CarteKiwi**
## The camera page is now available !
Use it to preview a camera device and to take a photo.
I tried to keep it simple. The whole camera process are in one single UserControl. 
Do not hesitate to give feedback to make it better :)

## PR checklist
- Which issue does this PR relate to? **[Issue 519](https://github.com/Microsoft/WindowsTemplateStudio/issues/519)**
- Anything that requires particular review or attention? **no**
- Do all automated tests pass? **can't execute tests, got this Catastrophic failure: [XUnit Issue 811](https://github.com/Microsoft/vstest/issues/811). Have to update package**
- Have automated tests been added for new features? **no**
- If you've changed the UI are you including screenshots to show the changes? **no changes, only one new page**
- Have you raised issues for any needed follow-on work? **no**
- Have docs been updated? **maybe I missed some**
- If breaking changes or different ways of doing things have been introduced, have they been communicated widely. **no breaking changes**

### CODE REVIEW

**Comment by crutkas** [comment]
i looked at this, i think we need to tweak some stuff and think how we can adjust for video.

**Comment by mrlacey** [comment]
A good start, but some comments:

- Not keen on the description "A page for capturing a photo from the camera device, include previewing, mirroring and orientation" poor grammar.
- Where's the French description from and why only that translation?
- `CircleButtonStyle` should go in a suitable style resource dictionary, not be in the page.
- The take photo button is assigned a Grid.Row value but is not in a grid with rows
- Don't use `#region`s 
- Take out the information only Debug calls from the generated code. They should not be needed
- Remove comments in the generated code that explain what the code is doing. (e.g. `// Initialize MediaCapture`)
- The switch statement on `device.EnclosureLocation.Panel` seems unnecessary for a single check. (default should be the last option as well)
- The long comment starting "// The call to stop the preview is included here for completeness, but can be" says that the code being included can be deleted. Why not just not include it? This is not a tutorial on how to use the API this should be production ready code. If necessary include a link to the docs in a comment.
- ` if (true)//_mirroringPreview)` ?!?
- Methods are inconsistently commented with XMLDoc tags. If they're really needed provide value for all sections. (Some are missing comments for the return value.)
- `ConvertDeviceOrientationToDegrees(SimpleOrientation orientation)` does not appear to be used.
- View needs to adjust for when window is very small in a NavView or hamburger icon crashes into page title
- Something's not right with the rotation support and image generation. Note the differences in the diagonal line in the image below.
image
- Support for the handling of no camera existing is inadequate. Giving no feedback is not sufficient.
- App crashes if have no camera but click on the "take photo" button
- Need a better explanation of the magic GUID `"C380465D-2271-428C-9B83-ECEA3B4A85C1"`
- "Take photo" button doesn't display correctly with dark theme
- Avoid hard coded colors (e.g. `BorderBrush="Black"`) as not theme compliant
- Should be using a Command, not a click event for event handlers in the MVVM version.
- Need to handle devices with front and back cameras and allow switching between them.
- In addition to showing the preview image, we also need documentation that explains how to access and use the captured image in other ways. The generated code should contain a comment that links to this doc. Documentation should also explain how to modify the size and format of the captured image.
- Generated code does not meet our styling and formatting requirements as defined by our StyleCop rules and checked in automated testing. Run the automated tests to see the full list of issues but here's sample output from one test:

  Views\CameraControl.xaml.cs(257,14): error SA1513: Closing brace must be followed by blank line 
  Views\CameraControl.xaml.cs(365,22): error SA1005: Single line comment must begin with a space. 
  Views\CameraControl.xaml.cs(34,140): error SA1028: Code must not contain trailing whitespace 
  Views\CameraControl.xaml.cs(5,1): error SA1209: Using alias directives must be placed after all using namespace directives. 
  Views\CameraControl.xaml.cs(281,21): error SA1013: Closing brace must be followed by a space. 
  Views\CameraControl.xaml.cs(281,22): error SA1013: Closing brace must be preceded by a space. 
  Views\CameraControl.xaml.cs(167,9): error SA1515: Single-line comment must be preceded by blank line 
  Views\CameraControl.xaml.cs(365,21): error SA1009: Closing parenthesis must be followed by a space. 
  Views\CameraControl.xaml.cs(36,50): error SA1214: Readonly fields must appear before non-readonly fields 
  Views\CameraControl.xaml.cs(25,9): error SA1124: Do not use regions 
  Views\CameraControl.xaml.cs(166,9): error SA1124: Do not use regions 
  Views\CameraControl.xaml.cs(291,9): error SA1124: Do not use regions 
  Views\CameraControl.xaml.cs(334,9): error SA1124: Do not use regions 
  Views\CameraControl.xaml.cs(278,60): error SA1500: Braces for multi-line statements must not share line 
  Views\CameraControl.xaml.cs(281,22): error SA1500: Braces for multi-line statements must not share line 
  Views\CameraControl.xaml.cs(281,21): error SA1500: Braces for multi-line statements must not share line 
  Views\CameraControl.xaml.cs(365,22): error SA1108: Block statements must not contain embedded comments [C:\UIT\SC07_2334\Proj\BlankCodeBehindAllStyleCop\BlankCodeBehindAllStyleCop\BlankCodeBehindAllStyleCop.csproj]

**Comment by CarteKiwi** [comment]
Thanks for these comments. I review each of them.
About the first two:

> Not keen on the description "A page for capturing a photo from the camera device, include previewing, mirroring and orientation" poor grammar.
> 
> Where's the French description from and why only that translation?

- What do you propose for the description?
- I'm French and I have translated from french to english. Please help me fill the description in english using rich grammar :)

**Comment by crutkas** [comment]
@CarteKiwi 

First, This is seriously a great start.

- For translation / wording, we can adjust that / create a bug after the fact and we can adjust this before it ships.
- @mrlacey did call out some critical issues.  Do you think you'll have time to address any?

**Comment by CarteKiwi** [comment]
@crutkas 

Thanks, I really appreciate.

- I have almost fixed each issues from @mrlacey . 

- This time, I'm making sure that StyleCop requirements are met.

- It remains me to add documentation to explain "how to modify the size and format of the captured image"

I should be able to push new commit in few hours.

**Comment by CarteKiwi** [comment]
Not complete yet. Productivity decreased due to vacation...
The switch command to switch between front and rear camera has not been tested on device (I missed my USB-C cable).

**Comment by crutkas** [comment]
@CarteKiwi did you see my comments?  anything we can do to help out?

**Comment by CarteKiwi** [comment]
@crutkas, I'm done with issues :)
The feature is ready.
As you said, if needed, we will adjust before it ships.

**Comment by mrlacey** [comment]
Looking over this again.

1. I get a file encoding test failure. But that should be easy to fix.
2. Includes French translations (@ralarcon - what's the impact of this for handling all translations? And, obviously, if we don't get everything else translated for the next release this needs reverting to English.--It'd be weird to ship with some (one) description in French and the rest in English)
3. "Package_postaction.appxmanifest" (both versions) have incorrect merge comment.
4. Adding "Styles\Button.xaml" could be done as part of Page.Camera.AddResource to remove the duplication.
5. The CompositionFilter for "Page.Camera.AddResource" doesn't need the framework check.
6. "Resources_postaction.resw" could be put in a composition template to remove the duplication. (We could probably also make this change in other templates too. It just stuck out at me because of the button file.)
7. "Package_postaction.appxmanifest" is also duplicated and so could move to composition.
8. "CameraControl.xaml" is duplicated and could move to composition.
9. "CameraControl.xaml.cs" is almost identical apart from some comments in one version. If the comments are no longer appropriate for one version they should be removed from both. And also can move to composition.
10. On CameraViewPage `SwitchBtn` and `ShootBtn` have inconsistent border thicknesses.
11. CameraViewPage.xaml shouldn't include the adaptivetriggers for adjusting the title. Instead, these should be added by modifying the filter of "Project.SplitView.Page.AddAdaptiveTriggerForPageTitle_Blank_Settings_TabbedPivot". That template could also probably do with a rename.
12. "MVVMLight.Page.Camera.AddCommandNS" seems unnecessary. just change the filter on "MVVMLight.Page.AddCommandNS".
13. The CodeBehind version of CameraViewPage.xaml.cs and CameraViewViewModel.cs contain hard-coded error messages that are not localized.
14. Improve grammar in the description. Change to "A page for capturing a photo from the camera. Includes handling previewing, mirroring, and orientation."
15. Update template.json description to "A page for capturing and displaying an image from the camera."
16. Need to test the output for accessibility support, etc. 
17. The forked branch wasn't synced with the latest VS15.3 support so I had challenges testing. If only someone would hurry up and make it easier to integrate support for more templates. [Oh wait](https://github.com/Microsoft/WindowsTemplateStudio/issues/919).
18. The page title is always the same and does not reflect the user specified page title. Highlighted items in image below should be the same
image
19. Clicking (or setting focus on and pressing) the take photo button many times in quick succession can lead to an unhandled COM error. There should be some basic error handling that stops the app from crashing and includes a note that action based on the error may need to be taken if just suppressing and ignoring the error has consequences.
20. I still don't understand why the PIP image is reversed from the live preview window. (see the diagonal line where the wall meets the ceiling.) That these are not the same seems like a bug to me.
image
21. Generated code still contains XMLDoc comments in a few places. The only other place we have these is in App.xaml.cs & the map LocationService. We should probably standardize this. Can ignore for now.
22. Repeatedly pressing the "Switch camera" button can cause the app to crash. As above, it shouldn't.
23.  I don't understand this comment or why it matters. "// TODO WTS: Set the file path and the name of the photo" I think this can be removed. It's definitely not a TODO.
24. Still missing
> documentation to explain "how to modify the size and format of the captured image"
25. Still look like a few StyleCop formatting issues.

Right now there are two things blocking me from being happy to take this in and then fix it up.
- Point 20 above
- The time necessary to address the above and then retest. (I won't have time this week.)

This has also prompted me to create #1037, #1039 & #1041 ;)

**Comment by CarteKiwi** [comment]
Thanks for your feedback.
Working on it.

**Comment by ralarcon** [comment]
@mrlacey @CarteKiwi, regarding the translations, initially, the localization files need to be in english for coherence until we start having translated content, so it is better to do not translate the content to any language. To include the language variation files is not required (we have an automated process to generate those the first time) but adding those is helpful.

We didn't have finished the exaclty process to follow to get the translations but, by now, will be handled internally (at least the first time). 

@CarteKiwi, you still have some work to do regarding code style and other tests not passing. Check the [build output](https://ci.appveyor.com/project/ralarcon/windowstemplatestudio-brpp6/build/MinIntegrationTests_0.14.16.dev-opdvhhys), it is not the best report to check for errors but going from bottom to top of the log you can see the tests exceptions (not so friendly -improvements comming). 

A better way to discover the issues is to run the tests locally. You can open the `Test.sln` solution and run, at least, the "Minimum" trait / set of tests. Whole integration tests run takes 1hour aprox and includes the build . The generation tests are recommended to be run before checking (those take about 10 mins and generate all the templates combinations).

**Comment by CarteKiwi** [comment]
@ralarcon Thanks. I previously said I was not able to run test because of **XUNIT issue 811**.
I updated the package locally, restarted VS and now I am able to run test.
This gonna change my build quality.
I also had another [runtime error](https://stackoverflow.com/questions/38671641/could-not-load-file-or-assembly-newtonsoft-json-version-9-0-0-0-culture-neutr) with the VSEmulator since the last merge on Dev: clicking on New C# project was crashing because of the NewtonJson package. 

Now, everything is OK, I can work properly.

**Comment by ralarcon** [comment]
@CarteKiwi... it was a long thread... :D Glad to hear you can now run the tests properly... I been doing some refactoring this week to help run the tests easily.

**Comment by crutkas** [comment]
@cartekiwi, AWESOME progress!

Can do the merge for v1.4 so we can then have more time to poke and adjust?  I think 1.3 may be a bit aggressive. 

I want to be sure we set expectations appropriately.

**Comment by CarteKiwi** [comment]
@crutkas Of course, it's wiser :)

**Comment by crutkas** [comment]
@CarteKiwi, i forked your repro and have a few hours to work on this while on an airplane.  I'll submit a PR to your repro so then you can PR in here :)

**Comment by CarteKiwi** [comment]
@crutkas Thanks. I will review this in few hours.

File: templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        {
+            _displayOrientation = sender.CurrentOrientation;
+
+            if (_isPreviewing)
+            {
+                await SetPreviewRotationAsync();
+            }
+        }
+
+        private SimpleOrientation GetCameraOrientation(DisplayInformation displayInformation, SimpleOrientation deviceOrientation)
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs]: Done

File: templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                        return SimpleOrientation.Rotated270DegreesCounterclockwise;
+                    case SimpleOrientation.Rotated270DegreesCounterclockwise:
+                        return SimpleOrientation.Rotated90DegreesCounterclockwise;
+                }
+            }
+
+            return result;
+        }
+
+        private int ConvertDisplayOrientationToDegrees(DisplayOrientations orientation)
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs]: Done

File: templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+                    return 180;
+                case DisplayOrientations.PortraitFlipped:
+                    return 270;
+                case DisplayOrientations.Landscape:
+                default:
+                    return 0;
+            }
+        }
+
+        private PhotoOrientation ConvertOrientationToPhotoOrientation(SimpleOrientation orientation)
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraControl.xaml.cs]: Done

File: templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
+<Page
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml]: Done for buttons

File: templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+          <views:CameraControl
+              x:Name="Camera"/>
+
+          <TextBlock 
+              Text="{x:Bind ErrorMessage, Mode=OneWay}"
+              FontSize="18"
+              HorizontalAlignment="Center"
+              VerticalAlignment="Center"/>
+
+          <RelativePanel
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml]: Done (removed)

File: templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+              Margin="20"
+              HorizontalAlignment="Center"
+              VerticalAlignment="Bottom">
+
+              <Button
+                  x:Name="SwitchBtn"
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml]: Done (removed)

File: templates/Pages/Camera/ViewModels/CameraViewViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            set { Set(ref _photo, value); }
+        }
+
+        public Panel Panel
+        {
+            get { return _panel; }
+            set { Set(ref _panel, value); }
+        }
+
+        public ICommand CaptureCommand { get; private set; }
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera/ViewModels/CameraViewViewModel.cs]: Done

File: templates/Pages/Camera.CodeBehind/Styles/Button.xaml [hunk 8]
```diff
@@ -0,0 +1,72 @@
+﻿<ResourceDictionary
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+    <Style x:Key="CircleButtonStyle" TargetType="Button">
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Styles/Button.xaml]: Done

File: templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        private async void Photo_Click(object sender, RoutedEventArgs e)
+        {
+            Photo.Source = new BitmapImage(new Uri(await Camera.TakePhotoAsync()));
+        }
+
+        private void SwitchCamera_Click(object sender, RoutedEventArgs e)
+        {
+            Camera.Panel = Camera.Panel == Windows.Devices.Enumeration.Panel.Front ? Windows.Devices.Enumeration.Panel.Back : Windows.Devices.Enumeration.Panel.Front;
```
[the remark this answers is not in the document]
Comment (CarteKiwi) [reply templates/Pages/Camera.CodeBehind/Views/CameraViewPage.xaml.cs]: Done

**Comment by crutkas** [comment]
@cartekiwi, i think a better route right now is lets create a file->new with WTS then only code review that sample so we can get it into a solid, awesome state.  From there, lets transition into templates.  code reviewing 80 files is hard with 3 separate frameworks.  I'm afraid a bug will be in one but not the others or something will be fixed in one but not the others with so many minor adjustments happening.

a lot easier to iterate and adjust that way versus working around the entire template engine.  This is how we do our initial work as well.

**Comment by mrlacey** [comment]
Styling (and position) of the switch button doesn't match that of the default camera app.
I would expect it to.
Why is it different?

**Comment by CarteKiwi** [comment]
You're right. I will change this.
By the way, it seems that the default camera app doesn't take in account the user theme.
Buttons looks exactly the same, in Dark and Light theme.
Do I ignore theme for these 2 buttons ?

**Comment by mrlacey** [comment]
>Buttons looks exactly the same, in Dark and Light theme.
>Do I ignore theme for these 2 buttons ?

No. Generated code should be theme and accessibility aware.

Default apps are a guide, not the last word.

**Comment by CarteKiwi** [comment]
@crutkas Are you telling me to focus myself only on one "template" (Code-Behind for example) ? You want me to create a shared single class hosting the "camera" code engine instead of a shared UserControl like the existing one ?
I am a little bit lost right now :/ 
I fixed almost every requested changes you made.
The current TODO list is the following :

> Add a page that shows capturing a photo from the camera (or library if no camera available) and displaying it on the page.

> CameraControl.xaml.cs should be abstracted out to a base library. The code would be heavily reused for a video page.

Am I wrong ? :)
If not, I will start working on these 2 tasks. It will take me some time to be done.

**Comment by ralarcon** [comment]
@CarteKiwi, did you progress in the requested changes? We are going to have a look on this to be added to the v1.4, so we will be adding some changes to align with the requested changes. We think we can work over your source branch ([CarteKiwi/WindowsTemplateStudio](https://github.com/CarteKiwi/WindowsTemplateStudio)) and to keep track of your excelent work, but we want to be sure we are not going to have collisions.

**Comment by CarteKiwi** [comment]
Hello @ralarcon, I did some changes, I will commit very soon.

**Comment by ralarcon** [comment]
Sounds good! Then we will review and if needed update directly over your branch thru a PR to allow you check the modifications. When ready we can merge this PR.

**Comment by dgomezc** [comment]
While testing an application generated in a windows 10 mobile, there is problem with the rotation:
- Taking a picture with the front camera, the preview image is rotated 90 degrees.
front_vertical

- Using the back camera and putting it in landscape mode, the preview of the capture seems to be in vertical mode and see two bars on the sides.
back_horizontal

The screenshots has been taken with the device emulator but also has been verified with a physical device.

**Comment by dgomezc** [comment]
@CarteKiwi, I will make a fork to your repository and make the following corrections:

- Extract the common camera code in CameraControl.xaml.cs and create a Camera Service.
- Replace calls to ResourceLoader and use the ResourceExtensions class.
- Rename Button.xaml style file to CameraPageStyles.xaml.

When finished, I will pull a request so you can merge the corrections.

**Comment by dgomezc** [comment]
Finally we aren´t going to create a CameraService since the logic is very coupled to the user control.

Moving the camera logic that is on the pages (buttons and control initialization) to the camera control itself.

Created dependency properties for customize button styles.

Improved accessibility.

**Comment by ralarcon** [comment]
@CarteKiwi, changes are ready, can you please review the PR we have done to your branch and solve the bugs mentioned above regarding the rotation?

We are waiting to include your contributions ;-)

**Comment by CarteKiwi** [comment]
@dgomezc thanks for your contribution :)
I am testing it, just to be sure everything is working as expected.
Then, I will solve the bugs regarding the rotation.

**Comment by ralarcon** [comment]
@CarteKiwi we are close now... but looks there should be an small conflict... can you please resolve it so we can merge the PR. Thanks! ;-)

**Comment by ralarcon** [comment]
Sorry @CarteKiwi, we were waiting to validate all and have the issues found be solved and in the mean while new conflicts has been appear. If you can solve the conficts I'll merge the PR asap and then we can fix the issues identified (#1219) as a separate thing.

**Comment by CarteKiwi** [comment]
@ralarcon I made the merge. 
After the PR is merged, I will have a look at #1219

**Pull request merged**